### ISSUE TITLE
merging of subgraph work into master

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#792]
So the `subgraph` and `subgraph_opts` branches have been sitting on the side for quite a while now, and there's a lot of interdependencies regarding getting those changes merged back into master, so here's an issue to discuss/track them.

First, these branches contain four general categories of changes:
1. basic Realm subgraph functionality and a unit test
2. lots of optimizations (copies, barriers, fusing, ...) performed on the Realm subgraph
3. Legion changes to turn traces into Realm subgraphs
4. performance optimizations related to Legion traces-as-subgraph replaying

Each of these has one or more preconditions for merging it into the master branch.  Items are listed below with checkboxes and owners.

Category 1:
- [ ] (@streichler) there's a bit of missing functionality in the Realm implementation (e.g. instantiating nested subgraphs) and the corresponding unit test
- [ ] (@streichler) we need some meaningful test coverage (e.g. the Legion/Regent CI tests all passing with tracing-generated subgraphs)

Category 2:
- [ ] (@streichler / @seemamirch ) the dma parts of these changes conflict pretty badly with other changes that have been made in master - either some heroic conflict resolution is needed or perhaps just re-implementation (e.g. by pasting chunks of code)

Category 3:
- [ ] (@lightsighter / @seemamirch ) these changes will conflict with changes in the control replication branch and we need a plan for what tracing with subgraphs will look like in CR before we merge into master

Category 4:
- [ ] (@lightsighter / @seemamirch ) determine whether these changes are desirable in their current form or should be implemented differently

Note that we do not need to fix everything before merging anything.  The actual dependencies are:
1->2, 1->3, 3->4

**Comment by streichler** [comment]
The changes for category 1 are now in the master branch with commit cd27595.

I was not able to rebase the category 3 stuff due to merge conflicts.  @seemamirch, you'll need to do the following:
```
git checkout subgraph_noopt
git rebase --onto origin/master 12de633
(then resolve all the conflicts)
```
If successful, the `subgraph_noopt` branch should then have exactly the set of changes needed for category 3 for you and @lightsighter to review.

**Comment by lightsighter** [comment]
Definitely not making this release, bumping to the next one.

**Comment by apryakhin** [comment]
We may need to push some optimizations for realm subgraphs in the context of legion tracing. Unclear the priority of this work at this time, so I will let @lightsighter @rohany to comment on the context . We seems to have some older history in this issue with the initial changes made to add subgraphs into realm but without any optimizations. From what I read we might have some of the "things" done already in `subgraph_opts` branch, so @streichler please correct me here. If we do have those already done but not merged probably worth discussing what they are and see if can dust off the branch and look into what's there and what else needs to be added/updated.

**Comment by lightsighter** [comment]
I suspect any changes in `subgraph_opts` have probably bit-rotted fairly badly compared with the latest branch of Realm and probably will not be usable anymore.

There is still a lot of work needed for Legion to be able to use Realm subgraphs which I will need to do at some point. I think I'm tentatively aiming to start that work early next year unless it becomes a priority sooner.

**Comment by rohany** [comment]
Something I'm working on with Joey is using a cheaper event representation within a subgraph to lower replay costs. For events that are internal to a single address space, the whole generational event structure is unnecessary, an atomic counter can suffice. We're looking at how much that buys us, and if more trimming of realm codepaths within subgraph replays are necessary.

**Comment by rohany** [comment]
Mike mentioned to me that something that also needs to be done in subgraph replay is the eager allocation of intermediate buffers for multi-hop copies. These buffers should be allocated upfront and passed to the copy engine during replay, rather than the copy engine dynamically allocating these buffers as copies are executed.

**Comment by muraj** [comment]
> These buffers should be allocated upfront

That may not be optimal.  IB buffers are limited in size, so holding an IB allocation for longer than a particular copy instance may cause other copies that need IB buffers space to fail.  There is no need to preallocate IB buffer space, as the memory is already allocated and registered, we just need to retrieve a free slab from the internal allocator.

**Comment by lightsighter** [comment]
> That may not be optimal. IB buffers are limited in size, so holding an IB allocation for longer than a particular copy instance may cause other copies that need IB buffers space to fail.

Indeed that is possible, but something similar is true with QQMDs in CUDA graphs. It's up to the user to make sure they don't make so many CUDA graphs that consume such resources for long periods of time. The same analogy should work for Realm graphs.

>  There is no need to preallocate IB buffer space, as the memory is already allocated and registered, we just need to retrieve a free range from the internal allocator.

True, but for the copies that do need intermediate buffers for multiple hops then we want those to run at SOL with Realm graphs, which means we don't want to pay for the active messages needed to reserve intermediate buffers dynamically when replaying graphs. All that should be happening when we replay a graph is for data to be moving, tasks to be executing, and atomic counters being decremented to signal what to do next. We shouldn't be doing any kind of dynamic resource management.

**Comment by muraj** [comment]
> Indeed that is possible, but something similar is true with QQMDs in CUDA graphs.

There's a couple of comments about CUDA graphs that I should mention, but most Ill mention offline as it goes into details (like about qmds) that likely shouldn't be discussed in a public forum.  First, copies like P2H2P or pageable copies don't pre-reserve staging memory to do their copies specifically because there's only one staging buffer per context and allocating from it would mean no other copy that needed it would function.  Second, cuda graphs doesn't do a lot of the optimizations it could or should for various reasons, so I wouldn't base my design off of how cuda graphs work too much.

> We don't want to pay for the active messages needed to reserve intermediate buffers dynamically

That's the thing, I don't think we need to.  Since we have the graph, we know when the earliest the copy can happen and can issue the copy from the correct node that makes the most sense rather than issuing a bunch of active messages to set things up remotely like we do in the normal path, removing the need to remotely allocate IB memory.

**Comment by eddy16112** [comment]
> Since we have the graph, we know when the earliest the copy can happen and can issue the copy from the correct node that makes the most sense rather than issuing a bunch of active messages to set things up remotely like we do in the normal path, removing the need to remotely allocate IB memory.

I think we are using push instead of get for our copies, so we need to send an active message to the receiver to allocate IB.

**Comment by lightsighter** [comment]
> First, copies like P2H2P or pageable copies don't pre-reserve staging memory to do their copies specifically because there's only one staging buffer per context and allocating from it would mean no other copy that needed it would function.

That seems reasonable for pageable memories, but I don't really care about them.

> Second, cuda graphs doesn't do a lot of the optimizations it could or should for various reasons, so I wouldn't base my design off of how cuda graphs work too much.

Also fine, but I'm hoping Realm graphs really try to push the limits of minimizing overheads even if it means consuming more resources.

> Since we have the graph, we know when the earliest the copy can happen and can issue the copy from the correct node that makes the most sense rather than issuing a bunch of active messages to set things up remotely like we do in the normal path, removing the need to remotely allocate IB memory.

I think regardless of which node issues the copy, there will need to be intermediate buffers allocated for some copies when they are multi-hop copies. The instances will remain the same from one replay to another and they will never be visible path between them. You'll always need an intermediate buffer to serve as the trampoline for the data to bounce of when it is moving regardless of who manages control of the data movement. 

As an aside, at least initially, most Legion traces will not be expressed in terms of a single Realm graph during multi-node execution. The reason for this is that if you replay a Realm graph today, that replay happens from exactly one node and that is not going to be scalable to dispatch work to a few thousand GPUs from one node. So until Realm graphs support "distributed replay", Legion traces in those settings are instead going to replay sharded Realm graphs, where each shard replays its own subset of the captured trace as a Realm graph. When there are dependences between sharded Realm graphs, we'll interpolate phase barriers to pass dependences between the Realm graphs in different shards. This means that you might not always be able to see the whole picture of a copy operation in a single Realm graph.

**Comment by lightsighter** [comment]
Yes, all the copies in the Realm graph should not have their own intermediate buffers. The Realm graph should do a "liveness" analysis when the graph is finalized (compiled) to determine the maximum number of copies that can potentially be running in parallel and then determine the minimum number and total size of intermediate buffers that are required to execute the graph without artificially serializing execution.
